### ISSUE TITLE
Eye bug (dunno what else to call it)

### ISSUE THREAD

**Issue by peardox** [castle-engine/castle-engine#342]
Refer to https://youtu.be/nq-_sU8vAlY for clear illustration of bug 

Comments from video explaining issue (repeated here for convenience - the middle one is MS's 3D Builder - the eyes look awful in that one)

This is a licensed model by a good artist on Sketchfab. I've got quite a lot of his stuff, he does good, expressive well animated work.

You'll notice that with the view3dscene playback the eyes of the model frequently glitch while this doesn't occur with two other players (so its a CGE thing...)

Of the artists work I have at least two other models that display exactly the same issue

Left it running for a whie so you don't need to keep on restarting it to see the problem.

Link to model = https://shrsl.com/35jnm - as it's Sketchfab you can view it in their online player (works there of course)

The model does raise three warning but they're only PBR
====================================================
3 warnings:

Material "Bear" has only PBR specular-glossiness parameters. We support it only partially (in particular specularGlossinessTexture is ignored). Better use metallic-roughness model.
Material "Eyes" has only PBR specular-glossiness parameters. We support it only partially (in particular specularGlossinessTexture is ignored). Better use metallic-roughness model.
Material "Eye_Shine" has only PBR specular-glossiness parameters. We support it only partially (in particular specularGlossinessTexture is ignored). Better use metallic-roughness model.

**Comment by michaliskambi** [comment]
@peardox Sorry it took so long to look into this.

I see the model is not free (as least right now) on Sketchfab, so I cannot (well, without paying) download the related gflTF. Are you OK with sharing the model with me privately? I am not going to use it for anything except to reproduce and fix this bug, of course.

I have a suspicion where's the bug (the eyes may reuse the same node in glTF, and in effect they reuse the same shape in X3D; this is good for memory, but also likely causes issue with the current way we do skinned animation in glTF). Still, fixing it "blindly" would be a challenge :) If I can look into the model it would be much easier.

You can e.g. send it as email to `michalis@castle-engine.io`. Thanks!

**Comment by michaliskambi** [comment]
@peardox Pinging about this. Is it possible (legally) to share the model with me? Or some other model that (likely) shows the same problem?

**Comment by peardox** [comment]
I'll ask the author for permission to share the model for investiatory use only

**Comment by peardox** [comment]
Emailed the authout now - CC'd your protonmail.com email

**Comment by peardox** [comment]
Model now available for investigation only legally - URL for download in Discord PM

**Comment by michaliskambi** [comment]
In progress. Here's a scary screenshot :)

I know it is not related to skinned animation, it's a regular problem of interpolating rotations. In some peculiar case our Slerp calculates something wild.

( This is good news in general, it means that the fix should be something local, straightforward. )

![bear_0](https://user-images.githubusercontent.com/1076978/220997746-07ffa895-4864-468d-a34a-683bf65defd3.png)

**Comment by peardox** [comment]
From the look of it I goessed it was a division close to (but not quite at) zero - same thing another way is the reciprocal of some number > 10k (think 10k is to do with IEEE 754 - had an issue with that specific one myself)

**Comment by michaliskambi** [comment]
https://user-images.githubusercontent.com/1076978/221064684-0f5b0251-f47e-462b-89a0-618fd3dc845a.mp4

1. Fixed!

    It was close to what you guessed -- we made a check at one point `"X > Epsilon"` to avoid division by zero in later `"... / X"`, but our Epsilon was a bit too large. The resulting calculation was... usually valid anyway, except in some corner cases.

    Fixed now, and I added automated testcase about it to `TTestCastleQuaternions.TestSlerpEpsilon`, so the issue will not return.

2. As a bonus, when I was playing with that code, I noticed an easy and valuable optimization for rotation interpolation. We had unnecessary quaternion -> axis-angle -> quaternion conversions.

   This isn't related in any way to the fix (the fix is valid, and necessary, both before and after this optimization).

As always, the fix is committed to latest CGE master, and will be available in CGE ( https://castle-engine.io/download ) and view3dscene snapshots ( https://castle-engine.io/view3dscene.php ) after 6-8 hours, when Jenkins will run all automatic tests and rebuilds the snapshots.
